### DISCUSSION TITLE
How to encrypt an ephemeral volume?

### DISCUSSION THREAD

**Discussion by schlichtanders** [kube-hetzner/terraform-hcloud-kube-hetzner#996] [Q&A]
Hi there,

I am looking for ways to encrypt everything at rest.
What I found so far:
- hetzner csi driver allows for encryption https://github.com/hetznercloud/csi-driver/blob/main/docs/kubernetes/README.md#volumes-encrypted-with-luks
- kubernetes allows to use any volume provisioner for ephemeral storage https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes

What I don't know:
- is hetzner csi the thing which accesses the plain local storage, like the storage which is used by default for ephemeral volumes? I really would like to reuse the storage of the hetzner instances and encrypt that

@mysticaltech can you help? Do you know someone who can help?

**Comment by schlichtanders** [comment]
I found it in this super list of csi drivers https://kubernetes-csi.github.io/docs/drivers.html

"A Container Storage Interface (CSI) Driver for Hetzner Cloud Volumes"

so this is not useful for encrypting the standard storage...

**Reply by schlichtanders** [reply]
@JustinGuese you have used hetzner csi encryption successfully, do you know whether it is also working with the default standard node storage?

Or is it really only used for accessing extra hetzner volumes?

**Reply by JustinGuese** [reply]
no it needs the hetzner csi as it is partly a function of the csi... but if you are running it on local volumes you could simply use Kasten K10s or OpenEBS and they will manage the local folder encryption and sync - so it's actually easier if you do not care about specific volumes (hetzner volumes)

**Reply by JustinGuese** [reply]
I would try OpenEBS if I were you, it's quite straightforward and should use local storage by default settings

**Reply by mysticaltech** [reply]
@schlichtanders I would follow @JustinGuese's advice. Also maybe @aleksasiriski can add something to that.

**Reply by aleksasiriski** [reply]
I'm using OpenEBS Localpath (alongside Longhorn, Hetzner Volumes and OpenEBS NFS with Hetzner Volumes). OpenEBS is the easiest to setup (with helm chart) but it's still ways behind Longhorn (although it looks very promising).

The only problem with OpenEBS is that it doesn't actually limit your volumes to the specified amount of storage that is requested, but to the amount of storage available on the node.

Here's the answer for [OpenEBS encryption at rest](https://openebs.io/docs/main/faqs/general#encryption-rest)

**Reply by schlichtanders** [reply]
thank you all for the hints!

actually it is not ideal for me to encrypt the entire node storage with one password.

The reason is that different actors may run code on the node, and I am afraid they somehow can get access to the underlying storage and reconstruct sensible data from other participants.

Hence I really would like to encrypt a single emptydir ephemeral volume with a given key.
I would be very interested whether OpenEBS or Kasten K10 can accomplish this - I haven't used them at all so far. Do they have respective csi drivers where I can set the key?

**Reply by JustinGuese** [reply]
hi @schlichtanders , if different actors can run code on the node encryption might not make a lot of difference anyways, because even if you would manage one encryption key per volume, they could most likely extract that encryption key from kubernetes itself :) except if you have fine-grained kubernetes authorisation set up

**Reply by schlichtanders** [reply]
thank Justin, of course I fine-grained kubernetes authorisation (this was the first I thought about). I also fine-grained network policies. Now the last piece missing is the encryption (as far as I understood Kubernetes).

**Reply by schlichtanders** [reply]
@aleksasiriski @JustinGuese I looked into the OpenEBS documentation, especially the [encryption at rest](https://openebs.io/docs/2.12.x/additional-info/faqs#encryption-rest)...
... it looks very complicated and even outdated as the two links about cStore and localPV point to an archived repository.

Do you have a resource about how you configure it in kube-hetzner so that it creates encrypted ephemeral volumes?


>The only problem with OpenEBS is that it doesn't actually limit your volumes to the specified amount of storage that is requested, but to the amount of storage available on the node.

@aleksasiriski this sounds crucial, as I need to make sure that no pod is eating the others temporary space ...

**Reply by schlichtanders** [reply]
I just realized that I also ideally need to encrypt the docker thin writable layer (which uses the overlay2 file system).

My final goal is to run different pods such that the one pod cannot extract sensible information from the other pod...
I still hope there is a somehow easy to maintain solution.

------

EDIT: I realized further that I actually only need to encrypt the docker writable layer (I realized it is actually containerd running inside k3s).

It seems this is not at all supported, but really everywhere people say you may encrypt the entire blockstorage (i.e. the entire node). So if multiple processes run on the same node (even with docker/containerd only), the processes may get access to the left-overs in the block storage of previous file-operations of other containers. Isn't docker/kubernetes also meant for isolation of different applications?

**Reply by schlichtanders** [reply]
I summarized my confusion into this discussion from containerd https://github.com/containerd/containerd/discussions/9148

I bet that my security concerns are not really apt in that if you only have a filesystem like overlayFS, you never can access the block storage directly and hence will never be able to read out other's sensible information.

If that is indeed true, the only thing left is to encrypt the entire node with LUKS (as far as I understood). (Not sure how to do that actually)

**Reply by schlichtanders** [reply]
I opened a new discussion for how to encrypt a node: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/997
